Source/Core/DolphinQt/Debugger/CodeDiffDialog.cpp

```diff
@@ -0,0 +1,406 @@
+  m_exclude_btn->setEnabled(false);
+  m_include_btn->setEnabled(false);
+  // Swap is used instead of clear for efficiency in the case of huge m_include/m_exclude
+  std::vector<Diff>().swap(m_include);
+  std::vector<Diff>().swap(m_exclude);
+  JitInterface::SetProfilingState(JitInterface::ProfilingState::Disabled);
+}
+
+void CodeDiffDialog::ClearBlockCache()
+{
+  Core::State old_state = Core::GetState();
+
+  if (old_state == Core::State::Running)
+    Core::SetState(Core::State::Paused);
+
+  JitInterface::ClearCache();
+
+  if (old_state == Core::State::Running)
+    Core::SetState(Core::State::Running);
+}
+
+void CodeDiffDialog::OnRecord(bool enabled)
+{
+  JitInterface::ProfilingState state;
+
+  if (enabled)
+  {
+    ClearBlockCache();
+    m_record_btn->setText(tr("Recording..."));
+    state = JitInterface::ProfilingState::Enabled;
+    m_exclude_btn->setEnabled(true);
+    m_include_btn->setEnabled(true);
+  }
+  else
+  {
+    ClearBlockCache();
+    m_record_btn->setText(tr("Start Recording"));
+    state = JitInterface::ProfilingState::Disabled;
+    m_exclude_btn->setEnabled(false);
+    m_include_btn->setEnabled(false);
+  }
+
+  m_record_btn->update();
+  JitInterface::SetProfilingState(state);
+}
+
+void CodeDiffDialog::OnInclude()
+{
+  const auto recorded_symbols = CalculateSymbolsFromProfile();
+  if (m_include.empty() && m_exclude.empty())
+  {
+    m_include = recorded_symbols;
+    return;
+  }
+
+  std::vector<Diff> current_diff;
+  if (m_exclude.empty())
+  {
+    current_diff = recorded_symbols;
+  }
+  else
+  {
+    for (auto& iter : recorded_symbols)
+    {
+      auto pos = lower_bound(m_exclude.begin(), m_exclude.end(), iter.symbol, AddrOP);
+
+      if (pos->symbol != iter.symbol)
+      {
+        current_diff.push_back(iter);
+      }
+    }
+  }
+
+  if (!m_include.empty())
+  {
+    m_include.erase(std::remove_if(m_include.begin(), m_include.end(),
+                                   [&](Diff& v) {
+                                     // Remove values if it can't find matching symbols
+                                     return std::none_of(
+                                         current_diff.begin(), current_diff.end(),
+                                         [&](Diff& p) { return p.symbol == v.symbol; });
+                                   }),
+                    m_include.end());
+  }
+  else
+  {
+    m_include = recorded_symbols;
+    RemoveMatchingSymbolsFromIncludes(m_exclude);
+  }
+}
+
+void CodeDiffDialog::OnExclude()
+{
+  const auto recorded_symbols = CalculateSymbolsFromProfile();
+  if (m_include.empty() && m_exclude.empty())
+  {
+    m_exclude = recorded_symbols;
+    return;
+  }
+
+  std::vector<Diff> current_diff;
+  if (m_exclude.empty())
+  {
+    m_exclude = recorded_symbols;
+  }
+  else
+  {
+    for (auto& iter : recorded_symbols)
+    {
+      auto pos = lower_bound(m_exclude.begin(), m_exclude.end(), iter.symbol, AddrOP);
+
+      if (pos->symbol != iter.symbol)

```

Comment:
```c++
if (pos == m_exclude.end() || pos->symbol != iter.symbol)
```

`std::lower_bound` can return `std::end()` (which shouldn't be dereferenced) if there is no lower bound. The new line between `current_diff.push_back(iter);` and `m_exclude.insert(pos, iter);` might also be removed.

Reply:
Done.
Removed the newline too.